Add reconfiguration for zigbee power meter to reduce reporting

varzac opened the pull request:
This PR adds code to reconfigure devices supporting the powerMeter capability to reduce the minimum reporting interval.  This changes to 5 seconds instead of 1 second.  On device init, if a device is detected as having not performed this migration (based on a field set on the device datastore), and supporting the capability a 5 minute timer is started if one isn't already running.  At the end of the 5 minutes, all devices the driver knows about are checked and the reconfiguration is attempted.  When we hear a success back from the device, the migration is marked as complete.  This will be attempted at most once per device init, but won't be reattempted periodically without another eligible device being initted.  This means for offline devices, this won't be re-attempted until the driver restarts.  I think in a future release I'd like to consider a more general solution in the firmware to handle future configuration changes.

cjswedes commented on drivers/SmartThings/zigbee-switch/src/configurations.lua:
```diff
@@ -60,8 +63,29 @@ local devices = {
+
 local configurations = {}
 
+configurations.power_reconfig_wrapper = function(orig_function)
```
Is orig_function supposed to be called from this function?

varzac replied:
lol, yes

cjswedes commented on drivers/SmartThings/zigbee-switch/src/configurations.lua:
```diff
@@ -60,8 +63,29 @@ local devices = {
+configurations.power_reconfig_wrapper = function(orig_function)
+  local new_init = function(driver, device)
+    local config_version = device:get_field(CONFIGURATION_VERSION_KEY)
+    if config_version == nil or config_version < driver.current_config_version then
```
I think we should also be using `Device:add_configured_attribute` to ensure if configure is called (I think some subdrivers do this from refresh handler) that the new configuration isn't overwritten.

varzac replied:
Done

varzac commented on drivers/SmartThings/zigbee-switch/src/configurations.lua:
```diff
@@ -60,8 +63,29 @@ local devices = {
+        end
+      end
+    end
+    device:set_field(CONFIGURATION_VERSION_KEY, driver.current_config_version, {persist = true})
```
Discussing with Kianoosh he wanted to get a quick change out with small change, and then review to provide a larger change later.  I think if we want to do a more complete look at the idea of "config versions" and checking for success, I would rather do that in the lua libs for the next release so it can be used in the future with any driver, and give us more time for review/testing.

greens commented on drivers/SmartThings/zigbee-switch/src/configurations.lua:
```diff
@@ -60,8 +63,29 @@ local devices = {
+      if device:supports_capability(capabilities.powerMeter) then
+        if device:supports_server_cluster(clusters.ElectricalMeasurement.ID) then
+          -- Increase minimum reporting interval to 5 seconds
+          device:send(clusters.ElectricalMeasurement.attributes.ActivePower:configure_reporting(device, 5, 600, 5))
```
I think that in the past we've run into race conditions when sending actual zigbee messages inside `init`.

varzac replied:
Thinking about this, the only issue I could see is if this is a child device (which is a real issue, and I will need to fix).  Do you have any more context on what the issues were?

greens replied:
there was a really old issue where the hub EUI occasionally wasn't set yet when some of these sends were called, iirc

it's possible that it has been fixed

varzac replied:
I re-worked this to schedule a timer to then do the actual configuration.

varzac commented:
Testing done:

Switched a single power meter to the driver, verified after 5 minutes new configuration was sent to the device.
Restarted driver, verified configuration does not happen again
Switched 2 more devices in quick succession, waited 5 minutes, verified both newly added devices were configured, but not the first device.

varzac commented:
> There is one failed test BTW that seems related to this and should be fixed.

Thanks, previously there was a matter test failing because of the capability stuff, and so I hadn't rechecked.  This was because the channel ordering needed to be relaxed since the order we send to the devices was not deterministic.  I've also added a test to cover the case that the device configuration is at 1, which was implicitly covered by the other tests, but I made an explicit one.

varzac commented:
Ok I have re-run all on target tests with the success status check as well as added several tests to cover different outcomes of the configure reporting response.  Technically there are some edge cases (specifically if there is a different configuration that happens while waiting for the response that is a success, but then this request is a failure), but they are unlikely enough, that I think that if any device hits it and stays on the older reporting, I consider that an acceptable outcome for a driver only addressing of this change.

tpmanley commented on drivers/SmartThings/zigbee-switch/src/configurations.lua:
```diff
@@ -60,8 +66,94 @@ local devices = {
+      device:set_field(CONFIGURATION_ATTEMPTED, true, {persist = true})
+    end
+  end
+  driver._reconfig_timer = nil
```
Just want to confirm I'm understanding correctly, the intention is to attempt the config at most once per device each time the driver starts up vs trying periodically in case the device comes back online?

varzac replied:
Yes, IMO the burden of running a perpetual timer to check all these things is more overhead than it's worth for the edge case of offline devices, and they will be covered the next time the driver restarts.
